File: leetcode/75.sort-colors.py
# @leet start
class Solution:
    def sortColors(self, nums: list[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        left = 0
        right = len(nums) - 1
        curr = 0

        while curr <= right:
            if nums[curr] == 0:
                nums[curr], nums[left] = nums[left], nums[curr]
                left += 1
                curr += 1
            elif nums[curr] == 1:
                curr += 1
            else:
                # 2
                nums[curr], nums[right] = nums[right], nums[curr]
                right -= 1


# @leet end

File: leetcode/75.sort-colors.ts
// @leet start
/**
 Do not return anything, modify nums in-place instead.
 */
function sortColors(nums: number[]): void {
  let left = 0;
  let curr = 0;
  let right = nums.length - 1;

  while (curr <= right) {
    if (nums[curr] === 2) {
      [nums[curr], nums[right]] = [nums[right], nums[curr]];
      right--;
    } else if (nums[curr] === 0) {
      [nums[left], nums[curr]] = [nums[curr], nums[left]];
      left++;
      curr++;
    } else {
      curr++;
    }
  }
}
// @leet end
